hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -523,17 +523,19 @@ object HoodieSparkSqlWriter {
     val params: mutable.Map[String, String] = collection.mutable.Map(parameters.toSeq: _*)
     params(HoodieWriteConfig.AVRO_SCHEMA_STRING.key) = schema.toString
     val writeConfig = DataSourceUtils.createHoodieConfig(schema.toString, path, tblName, mapAsJavaMap(params))
-    val bulkInsertPartitionerRows: BulkInsertPartitioner[Dataset[Row]] = if (populateMetaFields) {
+    val bulkInsertPartitionerRows: BulkInsertPartitioner[Dataset[Row]] = {
       val userDefinedBulkInsertPartitionerOpt = DataSourceUtils.createUserDefinedBulkInsertPartitionerWithRows(writeConfig)
```

Comment:
Here the client could use custom partitioner if they configured, also we have `RowCustomColumnsSortPartitioner` to let uses pass sort columns, this could also be enabled if `populateMetaFields` is false, right?

Reply:
Not really, `RowCustomColumnsSortPartitioner` again uses meta field `_hoodie_partition_path`, which is not present if `populateMetaFields` is false.

Reply:
The problem here is that the partitioner used here cannot rely on meta fields.  We can still allow user-defined bulk insert partitioner though, just that it should not be most of the build-in ones.

Reply:
Whether we should have a new method in `partitioner` to validate columns meet requirement(like return mandatoryFields, and we use it to check)? Currently if users set user-defined partitioner which acquire metafields, we will also accept it and not throw errors...